File: python/104_Maximum_Depth_of_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root == None:
            return 0
        self.maxDp = -float('inf')
        self.currentDepth = 0
        self.findMax(root)
        return self.maxDp
        
    def findMax(self,node):
        if node == None:
            return
        self.currentDepth += 1
        if node.left == None and node.right == None:
            self.maxDp = max(self.maxDp,self.currentDepth)
            return
        if node.left:
            self.findMax(node.left)
            self.currentDepth -= 1
        if node.right:
            self.findMax(node.right)
            self.currentDepth -= 1
        

File: c++/104_Maximum_Depth_of_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    int maxDep = 0;
    int currentDep = 0;
public:
    int maxDepth(TreeNode* root) {
        findMax(root);
        return maxDep;
    }
    
    void findMax(TreeNode* node){
        if(node == NULL){
            return;
        }
        currentDep++;
        if(node->left == NULL && node->right == NULL){
            maxDep = std::max(maxDep,currentDep);
        }
        else{
            if(node->left){
                findMax(node->left);
            }
            if(node->right){
                findMax(node->right);
            }
        }
        currentDep--;
    
    }
};
